primitives/src/pcs/mod.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) 2022 Espresso Systems (espressosys.com)
+// This file is part of the Jellyfish library.
+
+// You should have received a copy of the MIT License
+// along with the Jellyfish library. If not, see <https://mit-license.org/>.
+
+//! Polynomial Commitment Scheme
+pub mod errors;
+mod multilinear_kzg;
+pub mod prelude;
+mod structs;
+mod transcript;
+mod univariate_kzg;
+
+use ark_ec::PairingEngine;
+use ark_serialize::CanonicalSerialize;
+use ark_std::{rand::RngCore, vec::Vec};
+use errors::PCSError;
+
+/// This trait defines APIs for polynomial commitment schemes.
+/// Note that for our usage of PCS, we do not require the hiding property.
+pub trait PolynomialCommitmentScheme<E: PairingEngine> {
+    /// Prover parameters
+    type ProverParam;
+    /// Verifier parameters
+    type VerifierParam;
+    /// Structured reference string
+    type SRS;
+    /// Polynomial and its associated types
+    type Polynomial;
+    /// Polynomial input domain
+    type Point;
+    /// Polynomial Evaluation
+    type Evaluation;
+    /// Commitments
+    type Commitment: CanonicalSerialize;
+    /// Batch commitments
+    type BatchCommitment: CanonicalSerialize;
+    /// Proofs
+    type Proof;
+    /// Batch proofs
+    type BatchProof;
```

Comment:
Can add more trait bounds such as: 
`type Proof: Clone + Serialize + CanonicalSerialize + Debug + PartialEq...`

can take a look at [PairingEngine's associated type](https://docs.rs/ark-ec/0.3.0/ark_ec/trait.PairingEngine.html), they inherented a lot of basic traits.

Reply:
@mrain, @chancharles92  maybe use this?
```rust
    /// Prover parameters
    type ProverParam: Clone;
    /// Verifier parameters
    type VerifierParam: Clone + CanonicalSerialize;
    /// Structured reference string
    type SRS: Clone + Debug;
    /// Polynomial and its associated types
    type Polynomial: ark_poly::polynomial::Polynomial<E::Fr>;
    /// Polynomial input domain
    type Point: Clone + Copy + Debug + PartialEq + Eq;
    /// Polynomial Evaluation
    type Evaluation: Field;
    /// Commitments
    type Commitment: Clone + CanonicalSerialize + CanonicalDeserialize + Debug + PartialEq + Eq;
    /// Batch commitments
    type BatchCommitment: Clone + CanonicalSerialize + CanonicalDeserialize + Debug + PartialEq + Eq;
    /// Proofs
    type Proof: Clone + CanonicalSerialize + CanonicalDeserialize + Debug + PartialEq + Eq;
    /// Batch proofs
    type BatchProof: Clone + CanonicalSerialize + CanonicalDeserialize + Debug + PartialEq + Eq;
```

Reply:
- `type Polynomial` should be greatest common subset of traits bound from [`Polynomial`](https://docs.rs/ark-poly/0.3.0/ark_poly/polynomial/trait.Polynomial.html) and [`MultilinearExtension`](https://docs.rs/ark-poly/0.3.0/ark_poly/evaluations/multivariate/multilinear/trait.MultilinearExtension.html).
- `type Point` maybe exclude `Copy`

Reply:
Thanks @alxiong . But for `Polynomial`, `Rc` used in multilinear setting could not fit in this trait bound.
I think it make sense to put polynomial in heap because it's usually quite large.

Reply:
I agree, but similar to the argument for `impl Borrow<>`, how do you know what type of smart pointer one wants?

I don't think we should fix smart pointer type at the associate type level, (namely `MultilinearKzgPCS::Polynomial = Rc<DenseMultilinearExtension<E::Fr>>` is not as flexible as just `DenseMultilinearExtension<E::Fr>`. 

and you can always wrap this MLE into some smart pointer, such as `Rc`, `RefCell`, or `Arc` etc.

@chancharles92 what's your opinion?

Reply:
I don't have a strong opinion. Actually we were using `Polynomial = DenseMultilinearExtension`, but later Zhenfei changed it to `Polynomial = Rc<Dense...>` which has a few advantages:
1. It significantly simplifies the code, o/w we need to add Rc everywhere.
2. It make sure that the developer never forget to use `Rc`, otherwise the code performance can be accidentally effected.
3. When we want to use other smart pointers, we can e.g. use `Polynomial = Arc<...>` in associated type rather than change it everywhere.